File: logic/classes/Resolver.py
from classes.ErrorHandler import error_handler


class Resolver:
	def __init__(self):
		self.operators = ['!', '+', '|', '^']
		self.stack = []
		self.operands = []

	@staticmethod
	def catch_error(operands, values):
		if 'true' in values and 'false' in values:
			error_handler.throw_error('processing', 'Contradictory statements "{}"'.format(operands))

	@staticmethod
	def get_operator(implication, token, counter, operator):
		expected_operands = 1
		if token != '!':
			expected_operands = 2
		counter += expected_operands
		if not operator:
			operator = token
			implication = implication[:-1]
		else:
			counter -= 1
		return operator, implication, counter

	@staticmethod
	def get_operands(index, implication):
		left_operand = implication[:((index - 1) * -1)]
		rest = implication[((index - 1) * -1):]
		return [left_operand, rest]

	@staticmethod
	def get_query(operands, fact):
		result = []
		for operand in operands:
			if fact in operand:
				result.append(operand)
		return result

	@staticmethod
	def _not(operands):
		status = operands[0]
		if status == 'true':
			return 'false'
		if status == 'false' or status == 'none':
			return 'true'
		if status == 'undefined':
			return 'undefined'
		pass

	@staticmethod
	def _and(operands):
		if 'false' in operands:
			return 'false'
		if operands[0] == 'true' and operands[1] == 'true':
			return 'true'
		if 'none' in operands:
			return 'none'
		if 'undefined' in operands:
			return 'undefined'

	@staticmethod
	def _or(operands):
		if 'true' in operands:
			return 'true'
		if 'none' in operands:
			return 'none'
		if 'undefined' in operands:
			return 'undefined'
		return 'false'

	@staticmethod
	def _xor(operands):
		if 'true' in operands and ('false' in operands or 'none' in operands):
			return 'true'
		if 'none' in operands:
			return 'none'
		if 'undefined' in operands:
			return 'undefined'
		return 'false'

	def implication_not(self, operands, facts, fact, converted_flag):
		operands = filter(None, operands)
		if len(operands) > 1:
			error_handler.throw_error('processing', 'To many operands "{}"'.format(operands))
		if converted_flag:
			return self.evaluate_implication(operands[0], facts, fact, False)
		return self.evaluate_implication(operands[0], facts, fact, True)

	def implication_and(self, operands, facts, fact, converted_flag):
		if converted_flag:
			return self.implication_or(['{}!'.format(operands[0]), '{}!'.format(operands[1])], facts, fact, False)
		operands = self.get_query(operands, fact)
		values = [self.evaluate_implication(operand, facts, fact, False) for operand in operands]
		self.catch_error(operands, values)
		return reduce(min, values)

	def implication_or(self, operands, facts, fact, converted_flag):
		if converted_flag:
			return self.implication_and(['{}!'.format(operands[0]), '{}!'.format(operands[1])], facts, fact, False)
		queries = self.get_query(operands, fact)
		if len(queries) == 0:
			error_handler.throw_error('processing', 'Unknown fact')
		if len(queries) == 1:
			operands.remove(queries[0])
			if self.evaluate_condition(operands[0], facts) != 'false':
				if self.evaluate_condition(queries[0], facts) != 'false':
					return 'undefined'
				return 'none'
			return self.evaluate_implication(queries[0], facts, fact, False)
		if len(queries) == 2:
			if self.evaluate_condition(operands[0], facts) == 'false':
				return self.evaluate_implication(operands[1], facts, fact, False)
			if self.evaluate_condition(operands[1], facts) == 'false':
				return self.evaluate_implication(operands[0], facts, fact, False)
			return 'undefined'

	def implication_xor(self, operands, facts, fact, converted_flag):
		if converted_flag:
			left_condition = '{}!{}!+'.format(operands[0], operands[1])
			right_condition = '{}{}+'.format(operands[0], operands[1])
			return self.implication_or([left_condition, right_condition], facts, fact, False)
		left_condition = '{}!{}+'.format(operands[0], operands[1])
		right_condition = '{}{}!+'.format(operands[0], operands[1])
		return self.implication_or([left_condition, right_condition], facts, fact, False)

	def make_operation_condition(self, token, operands):
		if token == '!':
			return self._not(operands)
		if token == '+':
			return self._and(operands)
		if token == '|':
			return self._or(operands)
		if token == '^':
			return self._xor(operands)

	def make_operation_implication(self, token, operands, facts, fact, converted_flag):
		if token == '!':
			return self.implication_not(operands, facts, fact, converted_flag)
		if token == '+':
			return self.implication_and(operands, facts, fact, converted_flag)
		if token == '|':
			return self.implication_or(operands, facts, fact, converted_flag)
		if token == '^':
			return self.implication_xor(operands, facts, fact, converted_flag)

	def evaluate_condition(self, condition, facts):
		self.stack = []
		for token in condition:
			self.operands = []
			if token in self.operators:
				expected_operands = 1
				if token != '!':
					expected_operands = 2
				if len(self.stack) < expected_operands:
					error_handler.throw_error('processing', 'Not enough operands for operator')
				while expected_operands:
					self.operands.append(self.stack.pop())
					expected_operands -= 1
				self.stack.append(self.make_operation_condition(token, self.operands))
			else:
				self.stack.append(facts[token].resolve())
		if len(self.stack) == 1:
			return self.stack[0]
		else:
			error_handler.throw_error('processing', 'Condition error "{}"'.format(condition))

	def evaluate_implication(self, implication, facts, fact, converted_flag):
		counter = 0
		current_operator = ''
		for index, token in enumerate(''.join(reversed(implication))):
			if token in self.operators and counter != 1:
				(current_operator, implication, counter) = self.get_operator(implication, token, counter, current_operator)
			else:
				if counter == 0:
					if index != len(implication) - 1:
						error_handler.throw_error('processing', 'Implication error "{}"'.format(implication))
					if token == fact:
						if converted_flag:
							return 'false'
						return 'true'
					if token in facts:
						if converted_flag:
							return self._not(facts[token])
						return facts[token]
					else:
						error_handler.throw_error('processing', 'Unknown fact "{}"'.format(token))
				elif counter == 1:
					operands = self.get_operands(index, implication)
					return self.make_operation_implication(current_operator, operands, facts, fact, converted_flag)
				else:
					counter -= 1

File: logic/classes/Resolver.js
const { errorHandler } = require('./ErrorHandler');

class Resolver {
    constructor() {
        this.operators = ['!', '+', '|', '^'];
		this.stack = [];
		this.operands = [];
    }

    static catchError(operands, values) {
        if (values.includes('true') && values.includes('false')) {
            errorHandler.throwError('processing', `Contradictory statements ${operands}`);
        }
    }
    static getOperator(implication, token, counter, operator) {
        let expectedOperands = 1;
        if (token !== '!') {
            expectedOperands = 2;
        }
        counter += expectedOperands;
        if (!operator) {
            operator = token;
            implication = implication.slice(0, -1);
        } else {
            counter -= 1;
        }
        return [operator, implication, counter];
    }
    static getOperands(index, implication) {
        // TODO check slice
        const leftOperand = implication.slice(0, ((index - 1) * -1));
        const rest = implication.slice(((index - 1) * -1));
        return [leftOperand, rest];
    }
    static getQuery(operands, fact) {
        const result = [];
        for (const operand of operands) {
            if (operands.includes(fact)) {
                result.push(operand);
            }
        }
        return result;
    }
    static _not(operands) {
        const status = operands[0];
        if (status === 'true') {
            return 'false';
        }
        if (status === 'false' || status === 'none') {
            return 'true';
        }
        if (status === 'undefined') {
            return 'undefined';
        }
    }
    static _and(operands) {
        if (operands.includes('false')) {
            return 'false';
        }
        if (operands[0] === 'true' && operands[1] === 'true') {
            return 'true';
        }
        if (operands.includes('none')) {
            return 'none';
        }
        if (operands.includes('undefined')) {
            return 'undefined';
        }
    }
    static _or(operands) {
        if (operands.includes('true')) {
            return 'true';
        }
        if (operands.includes('none')) {
            return 'none';
        }
        if (operands.includes('undefined')) {
            return 'undefined';
        }
        return 'false';
    }
    static _xor(operands) {
        if (operands.includes('true') && (operands.includes('false') || operands.includes('none'))) {
            return 'true';
        }
        if (operands.includes('none')) {
            return 'none';
        }
        if (operands.includes('undefined')) {
            return 'undefined';
        }
        return 'false';
    }
    implicationNot(operands, facts, fact, convertedFlag) {
        // TODO check filter
        operands = operands.filter(item => item.length);
        if (operands.length) {
            errorHandler.throwError('processing', `To many operands ${operands}`);
        }
        return this.evaluateImplication(operands[0], facts, fact, !convertedFlag);
    }
    implicationAnd(operands, facts, fact, convertedFlag) {
        if (convertedFlag) {
            return this.implicationOr([`${operands[0]}!`, `${operands[1]}!`], facts, fact, false);
        }
        operands = this.constructor.getQuery(operands, fact);
        const values = operands.map(operand => this.evaluateImplication(operand, facts, fact, false));
        this.constructor.catchError(operands, values);
        return Math.min(...values); // TODO
    }
    implicationOr(operands, facts, fact, convertedFlag) {
        if (convertedFlag) {
            return this.implicationAnd([`${operands[0]}!`, `${operands[1]}!`], facts, fact, false);
        }
        const queries = this.constructor.getQuery(operands, fact);
        if (!queries.length) {
            errorHandler.throwError('processing', 'Unknown fact');
        }
        if (queries.length === 1) {
            operands = operands.filter(item => item !== queries[0]);
            if (this.evaluateCondition(operands[0], facts) !== 'false') {
                // TODO check this =>
                if (this.evaluateCondition(queries[0], facts) !== 'false') {
                    return 'undefined';
                }
                return 'none';
            }
            return this.evaluateImplication(queries[0], facts, fact, false);
        }
        if (queries.length === 2) {
            if (this.evaluateCondition(operands[0], facts) === 'false') {
                return this.evaluateImplication(operands[1], facts, fact, false);
            }
			if (this.evaluateCondition(operands[1], facts) === 'false') {
			    return this.evaluateImplication(operands[0], facts, fact, false);
            }
			return 'undefined'
        }
    }
    implicationXor(operands, facts, fact, convertedFlag) {
        if (convertedFlag) {
            const leftCondition = `${operands[0]}!${operands[1]}!+`;
			const rightCondition = `${operands[0]}${operands[1]}+`;
			return this.implicationOr([leftCondition, rightCondition], facts, fact, false)
        }
		const leftCondition = `${operands[0]}!${operands[1]}+`;
		const rightCondition = `${operands[0]}${operands[1]}!+`;
		return this.implicationOr([leftCondition, rightCondition], facts, fact, false)
    }
    makeOperationCondition(token, operands) {
        if (token === '!') {
            return this.constructor._not(operands);
        }
		if (token === '+') {
            return this.constructor._and(operands);
        }
		if (token === '|') {
            return this.constructor._or(operands);
        }
		if (token === '^') {
            return this.constructor._xor(operands);
        }
    }
    makeOperationImplication(token, operands, facts, fact, convertedFlag) {
        if (token === '!') {
            return this.implicationNot(operands, facts, fact, convertedFlag);
        }
		if (token === '+') {
            return this.implicationAnd(operands, facts, fact, convertedFlag);
        }
		if (token === '|') {
            return this.implicationOr(operands, facts, fact, convertedFlag);
        }
		if (token === '^') {
            return this.implicationXor(operands, facts, fact, convertedFlag);
        }
    }
    evaluateCondition(condition, facts) {
        this.stack = [];
        for (const token of condition) {
            this.operands = [];
			if (this.operators.includes(token)) {
			    let expectedOperands = 1;
				if (token !== '!') {
                    expectedOperands = 2;
                }
				if (this.stack.length < expectedOperands) {
                    errorHandler.throwError('processing', 'Not enough operands for operator');
                }
				while (expectedOperands) {
                    this.operands.push(this.stack.pop());
                    expectedOperands -= 1;
                }
				this.stack.push(this.makeOperationCondition(token, this.operands));
            } else {
                this.stack.push(facts[token].resolve()) // todo resolve  // todo what is facts
            }
        }
		if (this.stack.length === 1) {
		    return this.stack[0];
        }
		errorHandler.throwError('processing', `Condition error ${condition}`);
    }
    evaluateImplication(implication, facts, fact, convertedFlag) {
        // todo check carefully this function
        let counter = 0;
		let currentOperator = '';
		for (const [index, token] of implication.split().reverse()) {
		    if (this.operators.includes(token) && counter !== 1) {
		        const res = this.constructor.getOperator(implication, token, counter, currentOperator);
		        currentOperator = res[0];
		        implication = res[1];
		        counter = res[2];
            } else {
		        if (counter === 0) {
		            if (index !== implication.length - 1) {
		                errorHandler.throwError('processing', `Implication error ${implication}`);
                    }
		            if (token === fact) {
		                return !convertedFlag;
                    }
		            if (facts.includes(token)) {
		                if (convertedFlag) {
		                    return this.constructor._not(facts[token]); // todo what is facts
                        }
						return facts[token]; // todo what is facts
                    } else {
		                errorHandler.throwError('processing', `Unknown fact ${token}`);
                    }
                } else if (counter === 1) {
		            const operands = this.constructor.getOperands(index, implication);
					return this.makeOperationImplication(currentOperator, operands, facts, fact, convertedFlag);
                } else {
		            counter -= 1;
                }
            }
        }

    }
}

module.exports = {
  Resolver
};